Validate exclude patterns

lbausch opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------

Exclude patterns provided via `--exclude`, `--iexclude`, `--exclude-file` or `--iexclude-file` are validated. Invalid patterns cause `restic backup` to abort with a fatal error displaying the invalid patterns.
To achieve this, the `Pattern` struct is extended to include the original pattern. After parsing each pattern with `ParsePatterns`, every pattern part is validated using `filepath.Match`, by matching the pattern against itself.

Before Go 1.16 `filepath.Match` does not correctly report syntax errors, see https://go.dev/doc/go1.16#path/filepath
The corresponding tests are therefore only executed for Go 1.16+.

`restic restore` also accepts patterns - a follow-up MR could validate them as well.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

Closes #3709

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

greatroar commented:
> Some of the patterns used in tests, e.g. *.go are also invalid according to regexp.MustCompile().

That's because they're not regular expressions. They're extended glob patterns. No validation of those is possible other than checking for the empty string or an embedded zero byte, because practically all patterns are valid. They just won't match anything.

lbausch commented:
Ah, I see, so `regexp.MustCompile()` is not the right approach in this situation. Would it be sufficient to execute `filepath.Match()` with the pattern at hand against an empty string, e.g. `_, err := filepath.Match(patternStr, "")` and see if this raises any error (`err != nil`)?

fd0 commented:
> Would it be sufficient to execute `filepath.Match()` with the pattern at hand against an empty string, e.g. `_, err := filepath.Match(patternStr, "")` and see if this raises any error (`err != nil`)?

Yes, that should do it.

lbausch commented:
I've added some tests and they fail with Go 1.15. Apparently [1.16](https://go.dev/doc/go1.16) changed the way `filepath.Match` reports syntax errors:
>The [Match](https://go.dev/pkg/path/filepath#Match) and [Glob](https://go.dev/pkg/path/filepath#Glob) functions now return an error if the unmatched part of the pattern has a syntax error. Previously, the functions returned early on a failed match, and thus did not report any later syntax error in the pattern.

MichaelEischer commented on internal/filter/filter_test.go:
```diff
@@ -226,6 +228,49 @@ func TestChildMatch(t *testing.T) {
+		rtest.Assert(t, allValid == false,
+			"Expected invalid patterns to be detected")
+
+		rtest.Assert(t, reflect.DeepEqual(invalidPatterns, []string{"*[._]log[.-][0-9]"}),
```
`rtest.Equals` would make debugging the test easier

lbausch replied:
Changed it accordingly 👍🏻

MichaelEischer commented on internal/filter/filter.go:
```diff
@@ -216,6 +216,27 @@ func match(pattern Pattern, strs []string) (matched bool, err error) {
+func ValidatePatterns(patterns []string) (allValid bool, invalidPatterns []string) {
+	invalidPatterns = make([]string, 0)
+
+	for _, pattern := range patterns {
```
Wouldn't it be better to use `ParsePatterns` or `preparePattern` to preparse the pattern first?

lbausch replied:
Changed it accordingly 👍🏻

MichaelEischer commented on cmd/restic/cmd_backup.go:
```diff
@@ -298,6 +299,11 @@ func collectRejectByNameFuncs(opts BackupOptions, repo *repository.Repository, t
 			return nil, err
 		}
+
+		if valid, invalidPatterns := filter.ValidatePatterns(excludes); !valid {
```
The first two errors seem to be a duplicate for the last two?

lbausch replied:
I've tweaked the error messages to include the used flag, e.g. `--iexclude`

MichaelEischer commented:
> I've added some tests and they fail with Go 1.15. Apparently [1.16](https://go.dev/doc/go1.16) changed the way `filepath.Match` reports syntax errors

Go can build file conditionally on the compiler version, for example to require at least go 1.16, add as first line: `//go:build go1.16` . So let's just make that test conditional on a recent enough go version.

lbausch committed:
Keep original pattern for later use

lbausch committed:
Validate exclude patterns

lbausch committed:
Add tests for validating exclude patterns

lbausch committed:
Add changelog entry for validating exclude patterns

lbausch commented:
If I understand the usage of `//go:build go1.16` correctly, it would skip the entire file, so I used `t.Skip()` to just skip this specific test if running Go 1.15. Is this correct?

`ValidatePatterns()` now calls `ParsePatterns()` and tries to validate every single pattern part by matching it against itself. For being able to include the original invalid pattern in the error message show to the user, I've extended the `Pattern` struct to include the original pattern.

MichaelEischer commented:
> If I understand the usage of `//go:build go1.16` correctly, it would skip the entire file, so I used `t.Skip()` to just skip this specific test if running Go 1.15. Is this correct?

Yes, a `go:build` comment applies to the entire file. The usual solution as e.g. used by `internal/backend/rest/rest_int_go114_test.go` is to create a separate file for that specific test.

> Is a fatal error too much? One might argue that a backup including unwanted files is better than no backup at all.

Hmm, an invalid exclude pattern will currently cause restic to ignore all later excludes and backup too many files. Returning a fatal error would change that into not creating a backup at all. Actually just filtering out invalid patterns would also lead to a change of behavior. Returning a fatal error might be to cleanest solution although it could be dangerous for users which don't check the exit code of restic. @fd0 Do you have a preference what to do here?

fd0 commented:
Phew, that's a hard one. I'm (slightly) in favor of returning a fatal error. The pattern is clearly wrong and people **must** check the exit code.

lbausch commented:
Thank you :) I've updated the original MR description to describe all relevant changes.

MichaelEischer reviewed (approved):
LGTM. Thanks!

lbausch commented:
Thank you @MichaelEischer, @fd0 and @greatroar! 🙌🏻
